### PR TITLE
#5031, Kernel constructor cleanup

### PR DESCRIPTION
**Pull request by jonpsy**
Ding dong! , did anyone order cleanup? #5030 
@gf712

### CODE REVIEW

File: src/shogun/kernel/ANOVAKernel.h [hunk 1]
```diff
@@ -105,32 +86,21 @@ class ANOVAKernel: public DotKernel
...
-
-	/** register params */
-	void register_params();
 
 private:
 	float64_t compute_recursive1(float64_t* avec, float64_t* bvec, int32_t len);
 	float64_t compute_recursive2(float64_t* avec, float64_t* bvec, int32_t len);
 
 protected:
+	virtual float64_t compute(int32_t idx_a, int32_t idx_b);
```
Comment (gf712) [inline src/shogun/kernel/ANOVAKernel.h:96]: minor, but might be worth leaving compute where it was? So that the class members are separate from the class functions?

**Comment by gf712** [comment]
Thanks! Are you going to push more kernel refactors here or shall I merge when it passes CI?

File: src/shogun/kernel/ANOVAKernel.cpp [hunk 2]
```diff
@@ -7,44 +7,36 @@
...
 }
 
 ANOVAKernel::ANOVAKernel(
 	const std::shared_ptr<DenseFeatures<float64_t>>& l, const std::shared_ptr<DenseFeatures<float64_t>>& r, int32_t d, int32_t cache)
-  : DotKernel(cache), cardinality(d)
+  : ANOVAKernel(cache,d)
 {
-	register_params();
-	init(l, r);
+	ASSERT(DotKernel::init(l, r));
```
Comment (karlnapf) [inline src/shogun/kernel/ANOVAKernel.cpp]: I am curious, why is that changed?

File: src/shogun/kernel/ANOVAKernel.h [hunk 3]
```diff
@@ -105,32 +86,21 @@ class ANOVAKernel: public DotKernel
...
 	float64_t compute_rec2(int32_t idx_a, int32_t idx_b);
 
-	/** Casts the given kernel to CANOVAKernel.
-	 * @param kernel Kernel to cast. Must be CANOVAKernel. Might be NULL
-	 * @return casted CANOVAKernel object, NULL if input was NULL
+	/** Casts the given kernel to ANOVAKernel.
+	 * @param kernel Kernel to cast. Must be ANOVAKernel. Might be NULL
+	 * @return casted ANOVAKernel object, NULL if input was NULL
 	 */
 	static std::shared_ptr<ANOVAKernel> obtain_from_generic(const std::shared_ptr<Kernel>& kernel);
```
Comment (karlnapf) [inline src/shogun/kernel/ANOVAKernel.h]: I think these are not used anymore, so can be remove (see what happens if you do that)

File: src/shogun/kernel/AUCKernel.h [hunk 4]
```diff
@@ -54,66 +52,19 @@ namespace shogun
...
-			return "AUCKernel";
-		}
-
-		/** return feature class the kernel can deal with
-		 *
-		 * @return feature class SIMPLE
-		 */
-		virtual EFeatureClass get_feature_class()
-		{
-			return C_DENSE;
```
Comment (karlnapf) [inline src/shogun/kernel/AUCKernel.h:89]: removing the whole method? Does the superclass already have it?

File: src/shogun/kernel/ANOVAKernel.h (hunk 1 above)
Comment (jonpsy) [reply src/shogun/kernel/ANOVAKernel.h:96]: I see your point

File: src/shogun/kernel/AUCKernel.h (hunk 4 above)
Comment (jonpsy) [reply src/shogun/kernel/AUCKernel.h:89]: No. Scroll down a bit, I just used inline and made it into one line not removed

File: src/shogun/kernel/ANOVAKernel.cpp (hunk 2 above)
Comment (jonpsy) [reply src/shogun/kernel/ANOVAKernel.cpp]: Just removed init(l,r) and pasted its code here. Also DotKernel::init(l,r) returns a bool, so I asserted that.

**Comment by jonpsy** [comment]
> Thanks! Are you going to push more kernel refactors here or shall I merge when it passes CI?

There are a lot of kernels, some of them offer subtle complications. Also that I'm cleaning the docs too and the internal member names as well, so that'd be difficult for you guys to review(in case something bad goes down). I was instead planning on sending these in rounds.

EDIT: I'll add a few more simple cases, I'll @ you when I'm done ? :)  
EDIT2: Please merge? @gf712  :D

File: src/shogun/kernel/ANOVAKernel.h (hunk 3 above)
Comment (jonpsy) [reply src/shogun/kernel/ANOVAKernel.h]: Makes sense, removed it.

File: src/shogun/kernel/AUCKernel.h (hunk 4 above)
Comment (karlnapf) [reply src/shogun/kernel/AUCKernel.h:89]: ah I see, sorry :)

File: src/shogun/kernel/AUCKernel.h [hunk 5]
```diff
@@ -54,66 +52,20 @@ namespace shogun
...
 
-		/** return feature class the kernel can deal with
-		 *
-		 * @return feature class SIMPLE
-		 */
-		virtual EFeatureClass get_feature_class()
-		{
-			return C_DENSE;
-		}
+		inline virtual const char* get_name() const { return "AUCKernel"; }
```
Comment (karlnapf) [inline src/shogun/kernel/AUCKernel.h]: `inline` can be removed, it does nothing for us.
if you want to force use `SG_FORCE_INLINE` but I dont think that is needed here

Comment (gf712) [reply src/shogun/kernel/AUCKernel.h]: Yea, the compiler will do that for you since it’s such a small cost to inline it

File: src/shogun/kernel/LogKernel.h [hunk 6]
```diff
@@ -56,47 +56,23 @@ class LogKernel: public Kernel
...
-	 */
 	virtual float64_t compute(int32_t idx_a, int32_t idx_b);
 
-private:
-	void init();
-
 protected:
-	/// distance to be used
-	std::shared_ptr<Distance> m_distance;
+	std::shared_ptr<Distance> m_distance = NULL;
```
Comment (gf712) [inline src/shogun/kernel/LogKernel.h]: No need to set a shared ptr to NULL, the default constructor sets the pointer to nullptr

File: src/shogun/kernel/LogKernel.cpp [hunk 7]
```diff
@@ -11,46 +11,32 @@
...
-{
-	init();
-	ASSERT(m_distance)
-	
-	init(std::move(l), std::move(r));
+: LogKernel(10, degree, std::move(dist))
+{	
+	Kernel::init(l,r);
+	ASSERT(m_distance->init(l,r));
+	init_normalizer();
```
Comment (gf712) [inline src/shogun/kernel/LogKernel.cpp:34]: Where did this come from?

File: src/shogun/kernel/LogKernel.cpp [hunk 8]
```diff
@@ -11,46 +11,32 @@
...
-	
-}
-
-bool LogKernel::init(std::shared_ptr<Features> l, std::shared_ptr<Features> r)
-{
-	ASSERT(m_distance)
-	Kernel::init(l,r);
-	m_distance->init(l,r);
-	return init_normalizer();
-}
```
Comment (gf712) [inline src/shogun/kernel/LogKernel.cpp:48]: Deleting this is getting you the build error. This is implementing a pure virtual method from kernel

File: src/shogun/kernel/LogKernel.cpp (hunk 7 above)
Comment (jonpsy) [reply src/shogun/kernel/LogKernel.cpp:34]: removing its init(l,r) method (used only inside LogKernel constructor and pasted the code here.

Comment (gf712) [reply src/shogun/kernel/LogKernel.cpp:34]: Ok, but you need to keep the init(l,r) method so I guess just leave it there

Comment (jonpsy) [reply src/shogun/kernel/LogKernel.cpp:34]: Do we absolutely need to? I removed init(l,r) from other classes as well.

File: src/shogun/kernel/LogKernel.h (hunk 6 above)
Comment (jonpsy) [reply src/shogun/kernel/LogKernel.h]: I modified the default constructor to do only SG_ADD. So I guess I'd need to set default values to internal members no?

File: src/shogun/kernel/LogKernel.cpp (hunk 7 above)
Comment (gf712) [reply src/shogun/kernel/LogKernel.cpp:34]: Yes
https://github.com/shogun-toolbox/shogun/blob/7014676a95ec27b51473f037c4c137083075d61e/src/shogun/kernel/Kernel.h#L455

Comment (gf712) [reply src/shogun/kernel/LogKernel.cpp:34]: Actually it isn’t pure virtual so I guess you don’t need to

File: src/shogun/kernel/LogKernel.h (hunk 6 above)
Comment (gf712) [reply src/shogun/kernel/LogKernel.h]: No need for shared ptr, it will by default own a nullptr

File: src/shogun/kernel/LogKernel.cpp (hunk 7 above)
Comment (jonpsy) [reply src/shogun/kernel/LogKernel.cpp:34]: Very well, so I'll restore the init(l,r) methods. Maybe we can do something about them in future

File: src/shogun/kernel/LogKernel.h (hunk 6 above)
Comment (gf712) [reply src/shogun/kernel/LogKernel.h]: But primitive types require initialisation yes, otherwise it’s undefined behaviour

File: src/shogun/kernel/LogKernel.cpp (hunk 7 above)
Comment (jonpsy) [reply src/shogun/kernel/LogKernel.cpp:34]: > Actually it isn’t pure virtual so I guess you don’t need to

Yes but there are some methods (other than constructors) which are calling init(l,r).

Comment (gf712) [reply src/shogun/kernel/LogKernel.cpp:34]: Ah ok, then yes keep it

**Comment by gf712** [comment]
Thanks! :)
